### DISCUSSION TITLE
Error building nativeCompile

### DISCUSSION THREAD

**Discussion by christian2022** [AsamK/signal-cli#1453] [Q&A]
I'm trying to build signal-cli for Alpine using GraalVM and nativeCompile.
I get numerous errors in PosixDirectives.c with missing LC_ADDRESS, LC_IDENTIFICATION, LC_MEASUREMENT, LC_NAME, LC_NUMERIC, ...
Any idea?
```
$ docker run -it --rm --entrypoint /bin/bash ghcr.io/graalvm/native-image-community:21-muslib
bash-5.1# microdnf install git findutils
bash-5.1# git clone https://github.com/AsamK/signal-cli.git
bash-5.1# cd signal-cli/
bash-5.1# ./gradlew nativeCompile
...
[1/8] Initializing...                                                                                    (0.0s @ 0.18GB)
Error: Error compiling query code (in /tmp/SVM-5655773132912925684/PosixDirectives.c). Compiler command '/usr/local/musl/bin/gcc -Wall -Werror -o /tmp/SVM-5655773132912925684/PosixDirectives /tmp/SVM-5655773132912925684/PosixDirectives.c' output included error: /tmp/SVM-5655773132912925684/PosixDirectives.c:82:117: error: ?LC_ADDRESS? undeclared (first use in this function)
    method com.oracle.svm.core.posix.headers.Errno.ETIMEDOUT()
    C file contents around line 82:
    /tmp/SVM-5655773132912925684/PosixDirectives.c:81:     printf("NativeCodeInfo:PosixDirectives:ConstantInfo:ETIMEDOUT:PropertyInfo:value=%lX\n", ((unsigned long)ETIMEDOUT));
    /tmp/SVM-5655773132912925684/PosixDirectives.c:82:     printf("NativeCodeInfo:PosixDirectives:ConstantInfo:LC_ADDRESS:PropertyInfo:size=%lu\n", ((unsigned long)sizeof(LC_ADDRESS)));
    /tmp/SVM-5655773132912925684/PosixDirectives.c:83:     printf("NativeCodeInfo:PosixDirectives:ConstantInfo:LC_ADDRESS:PropertyInfo:signedness=$%s$\n", ((LC_ADDRESS>=0 ? 1 : 0)) ? "UNSIGNED" : "SIGNED");
...
```

**Comment by m-ueberall** [comment]
Judging from the `LC_` prefix without further looking into it, this _might_ be related to [musl libc locale limitations](https://wiki.musl-libc.org/open-issues.html#Locale_limitations).
However, your problems will unfortunately not end there. If you intend to build a (fully) static binary, you will also need a statically linked `libsignal_jni.a` which cannot be built without modification of the upstream sources ([see this comment](https://github.com/AsamK/signal-cli/issues/1305#issuecomment-1681917516), issue #1305).

**Reply by christian2022** [reply]
Thanks, yes that is what I'm trying to achieve. Or is there another way to make signal-cli run on Alpine?

You mentioned that you have to add `--static --libc=musl` to `native-image`. I only execute `gradlew nativeCompile`. Adding these parameters there doesn't work.

**Reply by morph027** [reply]
You might try https://wiki.alpinelinux.org/wiki/Running_glibc_programs#gcompat

**Comment by m-ueberall** [comment]
> Thanks, yes that is what I'm trying to achieve. Or is there another way to make signal-cli run on Alpine?

I haven't tried it myself yet, but since all existing (mostly) static binaries of `signal-cli` as well as the underlying dynamically linked library `libsignal_jni.so` (also needed when using a "normal" JRE in conjunction with the provided `libsignal-client-A.BB.C.jar` that contains it) require glibc, you might want to look into https://github.com/sgerrand/alpine-pkg-glibc which – from its description – should provide just that.
(In case the upstream build does not work because the above version of glibc is not sufficient/new enough, you could try a third-party build based on an older version of glibc–[I'm eating my own dog food here](https://media.projektzentrisch.de/temp/signal-cli/).)

> You mentioned that you have to add `--static --libc=musl` to `native-image`. I only execute `gradlew nativeCompile`. Adding these parameters there doesn't work.

The local build script injects those (and other) arguments by simply/blatantly renaming the `native-image` binary that comes with GraalVM and calling it through a dynamically rewritten wrapper that inherits its original filename (that is, "unbeknownst to Gradle"); something along the lines of …
```
#! /bin/env bash
echo ">>>> Executing /tmp/graalvm-community-openjdk-17.0.9+9.1/lib/svm/bin/native-image.org" -march=compatibility "$@"
exec "/tmp/graalvm-community-openjdk-17.0.9+9.1/lib/svm/bin/native-image.org" -march=compatibility "$@"
```
